Memory-efficient pull via pebble

Vultaire opened the pull request:
To allow for pulling larger files from pebble in a memory-efficient way, I wrote a custom parser which stores incoming file bodies into temporary files.  Other bits are still handled in memory as before.

Headers are still parsed using the same standard library logic.  However, handling of parts has been customized for our purposes.

This may not be the best solution, but maybe it can start discussion on how we want to handle this.

benhoyt commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,188 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+        match = self._boundary_rgx.search(self._buffer)
+        if match is None:
+            # Couldn't find a complete boundary in the buffer; stop for now
+            raise StopIteration()
```
Should this be `StopParsing`? If so, does this code path need to be tested?

benhoyt commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,188 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+        span = match.span()
+        data_before = self._buffer[:span[0]]
+        if data_before:
+            raise RuntimeError(
```
Would this (and similar ones below) be better as a `ProtocolError`? Because it's invalid data coming from Pebble, right?

benhoyt replied:
Ping on this.

benhoyt commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,188 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+            # bytes of the boundary line.
+            content = self._buffer[:-self._safe_buffer_len]
+            self._buffer = self._buffer[-self._safe_buffer_len:]
+            with self._open_tempfile() as outfile:
```
It seems inefficient to close and reopen the temp file every time we need to write to it (when we're only processing one file at a time). Can't we keep it open and just write to the open file?

benhoyt reviewed (changes requested):
This new state machine approach is really nice -- thanks for fixing this from a pure in-memory approach! One concern I have is that this is a fair bit of tricky parsing code and I want to make sure we've tested it thoroughly.

I realize there are existing push/pull tests that I wrote, but I think we should do a couple of things here:

1) Check the code coverage of the tests with a code coverage tool, to ensure we've tested all the edge and boundary cases. If some cases aren't tested, add new tests to cover those.
2) Add a few push/pull tests against the real Pebble server (see `TestRealPebble`).
3) For further thought: I wonder if we could re-use some tests or pull in some tests from another multipart parser to see how it compares to an existing one?

petevg commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,188 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+
+
+class StopParsing(Exception):   # noqa: N818
+    """Signals that the parser should stop parsing, at least for now."""
```
Something maybe feels off about triggering on the generic Exception class here, but I can't think why atm. I don't think that whatever is bouncing around in the back of my brain is super important, though.

Vultaire replied:
Yeah - from a pure FSM perspective, it may be arguable that we should transition to a paused state and track where things were, then transition back upon new data being served.  The exception seemed a simple way to avoid having to do that extra tracking.  I'd be more against it if the FSM in question were likely to grow more over time, but the scope here is relatively limited.

I'll try to come up with a way that avoids the exception; it may be easier than I could see at the time.

Vultaire replied:
OK, maybe this wasn't too bad.  Removed the exception, and also tried to remove multiple return paths at the cost of more explicit if/else logic.  I'm not sure that I fully like the latter part of that, but feedback (or proposed alternatives) are welcome.

petevg reviewed (commented):
I need to loop back and read this again. Things seem a little bit fussy right now, and the chain of dependencies doesn't make them nicer.

Nothing leaps out to me as incorrect (though I agree w/ some of Ben Hoyt's comments about Exceptions).

benhoyt reviewed (approved):
Thanks for this. Overall I'm happy with it, and am approving, though I would still like to see more done on the testing front.

Also, just the one comment I had left about `RuntimeError` vs `ProtocolError`.

jnsgruk commented:
Despite failing tests, I think this is ready for broader review while the minor test issues are worked through.

jameinel commented on ops/pebble.py:
```diff
@@ -1464,45 +1465,28 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
-            reader = io.BytesIO(content)
-        return reader
+        if len(filenames) > 1:
+            raise NotImplementedError('Multiple file responses not yet supported')
```
I wonder if this error could be a tiny bit clearer. You just did a 'pull(path)' there really shouldn't be any way to get >1 filename back, right?

niemeyer replied:
In addition to the comments from John, please note that the error messages are consistently lowercased, so that the typical "joining: like: this" works well visually.  Log messages are consistently titled-cased, though, for similar reasons (they are typically full sentences in the context in which they are presented).

pengale replied:
With the caveat that I haven't gone in to review what would happen right now if you threw something like `/home/ubuntu/somefiles/*.txt` at the pebble API, I kind of like this error.

In most cases, we will never reach this branch in the code. But one of the edge cases where we do hit it is when (and if) the Pebble API is updated to support a star or some other method of specifying multiple files in the path, and someone tries to be clever and just pass that through this function. In that case, "NotYetImplemented" is definitely the correct error to raise, as we do want to implement multi file support in the future. We just haven't done so yet!

(That said, we should lowercase the M)

rwcarlsen replied:
lower cased

jameinel commented on ops/pebble.py:
```diff
@@ -1464,45 +1465,28 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
-        # Then read the rest of the response and feed it to the parser.
+        parser = MultipartLargeFileParser(boundary)
         while True:
             chunk = response.read(8192)
```
It seems a bit of a shame that we have to buffer the entire content of the response into a local temporary file in order to read it. (you end up reading from disk, to write to a temp location on disk, to then read and put it somewhere else on disk).
It feels like 90% of the work has been done. All we would need to do is feed() until we got enough of the header information to know that we were reading the right content, and then return an object whose read() method returns either from the internal buffer, or calls feed() and then returns data.
That avoids tempfile and it avoids extra writes, and the in-memory buffer can stay reasonably bounded.

jameinel replied:
FWIW, I could be wrong on this. the json payload 'response' happens after the content if I remember correctly. Which may mean we have to buffer all of the content before we can understand what file we are processing. I think I'd still rather expose that a bit to the user, that they may have to buffer into a tempfile in their target directory and rename it into the final filename, rather than buffering potentially on another disk and having to copy it yet again.

I certainly wouldn't say that this isn't an improvement over what we have, but we are so tantalizingly close, I'd like to see if it is possible before we commit to this.

niemeyer replied:
Good points from John here.

rwcarlsen replied:
It is a bit of a shame.  However, as long as we preserve the API and prevent callers from seeing the file on-disk, we can preserve our ability to revisit this and optimize later as necessary.  As long as we can preserve future options, I'm okay with the single-path-putting-everything-on-disk (for now).

jameinel commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+            # Consume the header; now we're ready to parse the part body
+            # (either "request" or "files")
+            self._buffer = content
+        return next_state
```
I didn't get past this point for this review. I definitely like the design of the state machine, I feel like we are just really close to having a small buffer file object rather than buffering everything to temp.

rwcarlsen replied:
The FSM was dumped in favor of a stateful loop implementation.

jameinel commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+        self.current_filename = filename
+
+    def _get_open_tempfile(self):
+        return self._files[self.current_filename]
```
I don't actually understand how this works. In _prepare we do:
```
tf = NamedTemporaryFile()
tf.close()
self._files[name] = tf
```
So we store the object closed.
But `_get_open_tempfile` just returns the object without opening it.
And `_parse_files_part` does both:
```
outfile = self._get_open_tempfile()
outfile.write(content)
```
and
```
with self._get_open_tempfile() as outfile:
  outfile.write(content)
```
I'm guessing the 'with' will reopen the file? Though we also haven't directly defined the opening as 'append' mode, so if this is is reopening and writing, is it writing at the end?
I feel like there is a latent bug here, and we're missing a way to trigger it. We should make sure with coverage/etc that we are covering both paths through the _parse_files section.

pengale replied:
The context handler that with is calling is just self._get_open_tempfile, so there's no magic that will open the file. I think that we're both either missing something, or this is a good catch.

rwcarlsen replied:
`__enter__` for these tempfile objects (and TextIOWrapper, FileIO, etc. objects) does not re-open files.  Other than some validation/sanity checks, it just returns `self`.  So there is no double-opening issue here.  But it would be nice to get some tests that check both branches regardless.

jameinel commented on test/test_testing.py:
```diff
@@ -2751,6 +2751,19 @@ def _test_push_and_pull_data(self, original_data, encoding, stream_class):
+        client.push(self.prefix + '/test', original_data, encoding=None)
+        with client.pull(self.prefix + '/test', encoding=None) as infile:
+            received_data = infile.read()
+        self.assertEqual(original_data, received_data)
```
This certainly looks like we should be covering all the code paths, but it certainly doesn't make sense in *my* head why we have a file that may be open or closed at any given moment.

jameinel reviewed (commented):
Overall I like this a lot, but avoiding the extra buffering might require changing the signature (for example, maybe users can't expect the Response to have been read while their handling the streaming content).
I'd like us to explore that, and see what it looks like.

jameinel commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+        self._safe_buffer_len = len(boundary) + 6
+
+    def __del__(self):
+        for file in self._files.values():
```
I don't know if `__del__` acts quite as nicely as we would like it to.
Specifically, I think that `MultipartLargeFileParser` creates a reference counting loop with itself because it saves references to its methods, which means you have a refcount to a method which then has a refcount to the object.
```
>>> class Foo:
...   def __init__(self):
...     self._next = self._do_next
...   def _do_next(self):
...     print("next")
...   def __del__(self):
...     print("__del__")
...
>>> f = Foo()
>>> del f
>>> import gc
>>> gc.collect()
__del__
13
```
vs
```
>>> class Bar:
...   def __init__(self):
...     pass
...   def __del__(self):
...     print("Bar.__del__")
...
>>> b = Bar()
>>> del b
Bar.__del__
```

I don't know that there is a clean way to do this (at least __del__ isn't releasing the last references to MLFP, and MLFP doesn't have references to other objects, so I think it will trigger, but we *might* want a `MLFP.close()` which removes self._initial_state_fn.
(Though also all of this can be ignored if we move away from a tempfile we have to clean up.)

pengale replied:
I agree that this feels wrong. I would rather see an explicit call to clean up the temp file, if that's what we want to do, rather than relying on filesystem cleanup to happen as effectively a side effect of Python's garbage collector.

benhoyt replied:
Yep, it's typically to have an explicit `close()` method for things like this, or add `with` statement support by defining `__enter__` and `__exit__` (though the latter might be a bit heavy here).

rwcarlsen replied:
There are actually two issues here: closing the file, and removing the temporary file.  In order to facilitate the user getting an open file with their own mode/encoding params different from those used when the parser is writing the file, we need to close+re-open before we hand the object/file back to the caller of `pull`.  Because of this, the tempfile's auto-delete-on-close has been disabled.  If we want the charm-container to behave a bit more stateless by default, then it would be nice to auto-remove this temporary file upon close.  And the caller would be free to copy the data into a non-temp location if they want to persist it.  But if the "normal" use case is to have the file persist, then we should just kill this `__del__` code.  To accomplish the former - we basically would need to use something like the tempfile lib does internally with its `_TemporaryFileWrapper` class.  I'm new enough to this project that I'm not sure which behavior is more natural/preferrable.

jameinel replied:
I don't think we want our copy of the content buffer to exist, as it will never be in the right location for the caller.

rwcarlsen replied:
Gustavo made a good suggestion (which I've implemented).  Before we return the (opened) file to the caller of `pull`, we can remove the file from disk.  The data will remain available until the file is closed (by the user or the gc).  This way we don't introduce any new API (e.g. a file on disk the user can do stuff with) - and limit them to reading from returned object directly.  We can continue to ruminate on temp file locations, copying, etc. details until we are confident if/how we want to change things.  I'm going to consider this resolved for now.

niemeyer reviewed (changes requested):
Thanks a lot for looking into this. A few initial comments:

Vultaire commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+        # boundary line by mistake.  This is easy if we detect the boundary in full, but if we
+        # haven't received it yet, we need to be careful not to feed any partial boundary line
+        # to disk.  This minimum buffer length threshold will ensure that, and will work for both
+        # regular and terminating boundary cases.
```
Hi Gustavo - thanks for taking the time to look at this.

I'm back with the BootStack team right now, but I will try to review your comments as soon as time reasonably allows.

I did want to highlight one detail regarding the implementation, and that is because of how the data is served via requests.  While in the HTTP datastream itself it is totally correct that things start with a CRLF, the data served from requests does not include that CRLF.

This highlights what I'm talking about:
* REPL session using ops.pebble.Client._requests_raw(): https://pastebin.ubuntu.com/p/XVHTjhSycK/
* socat output showing data transferred via the socket: https://pastebin.ubuntu.com/p/PQsD26nCsc/

Again, I'll take another look at this when I have some time to properly read and reflect on all the feedback you've provided, but I did want to point out the above detail as I distinctly remembered it as a quirk that came up when implementing this.

niemeyer replied:
Thanks for highlighting this. That could mean we need to special-handle the very first boundary indeed. At the same time, it doesn't seem to change the remaining points, in the sense that inside that response, that's still how it generally works, and we should not blindly assume that \r\n is there without verification.

Vultaire commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+            span = match.span()
+            # Content is terminated by \r\n<boundary>; adjust for the \r\n
+            end_of_content = span[0] - 2
+            content = self._buffer[:end_of_content]
```
I've got a chance to review all of your comments.  I'm not sure how soon time will allow me to address them in code, but I'd like to at least reply more completely.  I'll keep it brief though.

* Understood re: error/log strings and lower- vs. title-casing.  I'll make the changes.

* I think you have a good point re: the CRLF at the beginning of the boundary.  If that were included, I think that would eliminate the need for the ugly "-2" index adjustments.  As I mentioned before, the initial CRLF appears to not be included in the request body by the requests library, i.e. it's removed as part of header parsing.

  If I were to prime the initial buffer with "\r\n" to reflect that bit, that may be a smaller wart and would allow for the rest of the logic to be more clean.

* Re: regex vs. substring matching for the boundary line: this is because there's 2 types of boundary lines and I was trying to allow for one pattern to match both.  Specifically: ```<CRLF>--<boundary_str><CRLF>``` in the normal case, and ```<CRLF>--<boundary_str>--<CRLF>``` in the terminal case.

  Because the "--" suffix for terminal strings is stuck in the middle of the pattern prior to the final CRLF, it felt reasonable to use a regex for that so I could simply detect a complete boundary without having to use 2 distinct strings.  However, you also had a good point; I did nothing to escape the boundary value.  So clearly, something should change there.

* Re: FSM or no: I'm open to alternatives.  I'd like to perform some reworks for the other items noted here first, but if that is not sufficient (or if you strongly disagree), I'm open to discussing this.

* Re: how things get fed into the buffer: I combine the remainder of the last chunk with the new chunk to ease matching for substrings/regexes.  That's the simple reason.  The buffer is continually trimmed throughout the stages of this FSM, and chunk sizes should not be terribly large, thus that's why I felt that was appropriate.  I am very happy to discuss this further though.

petevg commented:
The tests on this one are failing because the regex pattern for our boundary is ```boundary_rgx_pattern = r"--{}(--)?\r\n``` (windows newlines), and the boundary strings we are submitting in the tests have unix newlines. I'm not sure which is correct. (Possibly, we should be checking for both?)

jameinel commented:
I wonder if we are reading a file in text mode vs binary mode. IIRC HTTP
transfer protocols actually specify that it *must* be exactly `\r\n`
(carriage return and line feed). Not just one or the other especially for
headers and boundaries. That sounds like something is genuinely wrong on
Windows.


On Fri, Dec 17, 2021 at 7:31 PM Pen Vander Giessen ***@***.***>
wrote:

> The tests on this one are failing because the regex pattern for our
> boundary is boundary_rgx_pattern = r"--{}(--)?\r\n (windows newlines),
> and the boundary strings we are submitting in the tests have unix newlines.
> I'm not sure which is correct. (Possibly, we should be checking for both?)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/operator/pull/667#issuecomment-997106657>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRQ7I6FJTNO7W4PHA26GTURPI4PANCNFSM5HTDK7OA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

Vultaire commented:
I haven't been able to take a look at this but just noticed the comments.  \r\n is correct at the protocol level.  It sounds like an issue with reading a file in text mode on Windows.  Probably the test in question needs to test using binary data to avoid the conversion of \r\n to \n on Windows.

Vultaire commented:
It looks like one of my latter changesets (changeset 6dc39a82c34e06a4f91d15bd097ebc0859032dbb, "Rework: leave tempfiles open until completion of writing data") broke the "real" pebble tests used for validating this.  I'll try to find time to repair this.

Vultaire commented:
I just did a correction to one of my patches to fix some of the test errors, and am looking at the remaining issues which appear to be coming from test.test_pebble.

Specifically, I'm seeing failures in tests from test.test_pebble.TestClient.  After reviewing the tests, my interpretation is that I'm surprised the tests worked to begin with.  These are tests which leverage the MockClient, which inherits from the ops.pebble.Client modified by this MR.  A number of these tests now fail because of the new code.

Upon looking at the tests, I see multiple issues:
1. We have multi-line bytes literals representing response bodies.  These response bodies do not include carriage returns when checked out with UNIX-style line endings.  This appears to be something that the previous parser was forgiving about, but which is technically incorrect per RFC 2046, which expects CRLF line endings on the boundaries rather than merely LF.
2. The response bodies also include a leading LF.  This does not match the behavior of the requests library; it parses the request headers and strips the particular CRLF, starting the request body on the boundary line.

My next patches will try to correct these tests in a way which should also work if the sources are checked out with Windows-style line endings.

Vultaire commented:
Note: PEP 278 seems to accomodate for Windows vs. UNIX line endings in multiline constants, treating both as LF.

pengale commented:
> I just did a correction to one of my patches to fix some of the test errors, and am looking at the remaining issues which appear to be coming from test.test_pebble.

Thank you for continuing to look into this! I know that you're busy, and I really appreciate not having to catch up on some of the gotchas here right now.

Let me know if you run out of time, and need to hand things off, though ...

Vultaire commented:
I think I've fixed the unit test breakage overall, although I'm not certain about Windows at this point.

There is more rework I wish to do re: Gustavo's points, but for the time being I'm going to have to step away again.  Others are welcome to pick up the torch.  I made some comments in my reply to Gustavo; I also see some bits which seem less readable than I previously thought when I went back to review the code over these last few days, so perhaps some cleanup and refactoring is also in order.

benhoyt commented on test/test_pebble.py:
```diff
@@ -1481,21 +1481,21 @@ def test_get_services_names(self):
+            b"""\
+--01234567890123456789012345678901\r
+Content-Disposition: form-data; name="files"; filename="/etc/hosts"\r
+\r
```
I don't understand how these `\r` and suchlike are working. Isn't `b"""` a raw string, meaning that there are no escape sequences, and `\r` will end up as a literal backslash followed by a literal `r`?

benhoyt replied:
Aha, @jameinel's right, sorry. `b` is byte, `r` is raw, and `"""` just means multiline (but still processes escapes). I had to type it into a Python REPL to re-convince myself. :-) I've been doing too much Go, where backtick strings are multiline *and* raw.

jameinel commented:
b is a byte string, not a raw string.
r'' is a raw string, b is just saying it shouldn't be treated as unicode
chars.


On Wed, Jan 5, 2022 at 4:09 PM Ben Hoyt ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/test_pebble.py
> <https://github.com/canonical/operator/pull/667#discussion_r779138203>:
>
> > @@ -1481,21 +1481,21 @@ def test_get_services_names(self):
>      def test_pull_text(self):
>          self.client.responses.append((
>              {'Content-Type': 'multipart/form-data; boundary=01234567890123456789012345678901'},
> -            b"""
> ---01234567890123456789012345678901
> -Content-Disposition: form-data; name="files"; filename="/etc/hosts"
> -
> -127.0.0.1 localhost  # """ + b'\xf0\x9f\x98\x80\nfoo\r\nbar' + b"""
> ---01234567890123456789012345678901
> -Content-Disposition: form-data; name="response"
> -
> +            b"""\
> +--01234567890123456789012345678901\r
> +Content-Disposition: form-data; name="files"; filename="/etc/hosts"\r
> +\r
>
> I don't understand how these \r and suchlike are working. Isn't b""" a
> raw string, meaning that there are no escape sequences, and \r will end
> up as a literal backslash followed by a literal r?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/operator/pull/667#pullrequestreview-845066103>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRQ7MMW2D63ZIV2ZUDPBDUUSXQ3ANCNFSM5HTDK7OA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

Vultaire commented:
Summarizing today's changes:
* Cleaned up a couple of tests so they don't leave dangling file handles.
* Primed buffer with CRLF for the special initial case re: boundary detection, letting us include the initial CRLF in the boundary pattern.
* Replaced boundary regex with 2 distinct byte strings, to avoid unexpected issues with escapes and regex.
* Extracted some code into helper functions for reuse or clarity.
* Ultimately moved away from the FSM model in favor of a more straightforward handler function.

Changes were performed against existing unit tests (including the "real pebble" tests); no new tests were added.

pengale commented on ops/pebble.py:
```diff
@@ -1926,3 +1910,198 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer += data
```
I think that, if we wanted to rely on automagic in Python to be reasonably efficient here, we'd want to use `.join` rather than the `+` operator. That may be dated knowledge -- I'm returning to focusing on Python after a break, myself.

Regardless, that would not address @niemeyer's concerns, because we'd still be making a big, redundant copy of the stuff in memory. Python would just be able to do more of the allocation of the extra memory in advance.

Refactoring to use `self._buffer = data` would be more efficient, because we'd just be changing a reference. That would discard any data left in the buffer, though, which feels intuitively wrong (though might be okay in this case).

This makes me want to set aside some time to build some instrumentation around memory use and such in OF. Without concrete data, it's hard to say how much of an issue the copy is. We are working with big files, but we are working with them in http sized chunks, which is less concerning to me than if we were reading them directly off a local disk.

benhoyt replied:
Could we use a `bytearray()` and `self._buffer.extend(data)` to avoid an extra allocation and copy? That way we only need to copy data down when we have a partial buffer, which should only happen once at the start of a large file.

rwcarlsen replied:
Ben is right - we should be able to avoid the reallocation issue by extending a mutable type (e.g. bytearray) rather than using the immutable bytes string.  Doing `self._buffer = bytearray()` in the parser `__init__`  and calling `extend` should do the trick.

pengale replied:
@rwcarlsen thank you for the comment (and thank you @benhoyt for the other comment, which I managed to miss).

I agree that using a bytearray() is the right thing to do here, with no extra performance testing needed. (This doesn't change my desire to drop a performance testing harness into OF; it just shouldn't block progress on this PR.)

pengale reviewed (commented):
I agree that this feels close. I think that there are two outstanding issues:

1) Look at how we delete the tempfile, and possibly change how we do so. (Since we're deleting on GC, I'm not clear why we're passing delete=False into the routines in tempfile when we create the file.)

2) Think about memory efficiency wrt dropping the raw bytes into our buffer. I'm tempted to suggest that we not worry about this for now, but instead make plans to build some tooling around memory, network, and cpu efficiency, and refactor the code based on data. This might be an easy (ish) win, though, so it deserves some thought.

Also, I just want to say that I am really grateful for all the work that @Vultaire has done here, even after he officially had to switch to other duties. Those two final points are nice and focused, and much less of a headache than trying to wade in and tackle carriage returns and mysteriously failing tests :-)

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
         self._request('POST', '/v1/signals', body=body)
+
+
+class MultipartLargeFileParser:
```
Nit: this is used for all files, right, not just large ones? What about removing `Large` from the name?

rwcarlsen replied:
Renamed.

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        # and the optional "--" suffix.  Pebble doesn't use this, so to simplify implementation I
+        # am ignoring it.
+        self._boundary = "\r\n--{}\r\n".format(boundary).encode()
+        self._terminal_boundary = "\r\n--{}--\r\n".format(boundary).encode()
```
I believe this is incorrect, as `'{}'.format(my_bytes)` will give include the `b'` quotes in the output. (It would probably still work, but not be what we want. Also, just concatenating the bytes avoids the call to `format` and `encode`:

```
>>> '\r\n--{}\r\n'.format(boundary).encode()
b"\r\n--b'0123456789abcdef'\r\n"
>>> b'\r\n--' + boundary + b'\r\n'
b'\r\n--0123456789abcdef\r\n'
```

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer += data
+        # self._run_parser_fsm()
```
Nit: this commented-out line should be removed.

rwcarlsen replied:
removed.

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        while True:
+            # Try to read headers, if present.
+            if self._headers is None:
+                if self._buffer.startswith(self._boundary) and header_terminator in self._buffer:
```
Do we need to skip optional whitespace before the initial boundary? I don't feel good about how we've removed the whitespace from the start of the test cases...

rwcarlsen replied:
Indeed - I found this in the spec:

```
There appears to be room for additional information prior to the
first boundary delimiter line and following the final boundary
delimiter line.  These areas should generally be left blank, and
implementations must ignore anything that appears before the first
boundary delimiter line or after the last one.
```

rwcarlsen replied:
I updated the code to handle this, but the tests still need to not have the leading whitespace unless we re-introduce the `\r\n` prefix to the boundary marker (this is related to the issue where the python lib eats that marker before our parser sees the response data).

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    data = self._buffer[:next_boundary_index]
+                    self._response = json.loads(data.decode())
```
`json.loads()` takes a str or bytes or bytearray, so I don't think you need `.decode()` here.

rwcarlsen replied:
Turns out that json in python 3.5 only takes `str`.  So this needs to stay like it is.

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                if next_boundary_index is not None:
+                    # Next boundary's location is known; this is the final write.
+                    data = self._buffer[:next_boundary_index]
+                    with self._get_open_tempfile() as outfile:
```
It seems clearer to use an explicit `outfile.close()` here rather than a `with` statement, as `with` is normally used when you're opening and closing the file. So:

```
outfile = self._get_open_tempfile()
outfile.write(data)
outfile.close()
```

rwcarlsen replied:
done

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to a boundary line.
+                    safe_bound = len(self._buffer) - len(self._terminal_boundary)
```
Nice -- this cross-boundary edge case would be easy to miss. :-) Do we have an explicit regression test for the boundary crossing the chunk size? I think we should if not.

Speaking of which, I think we should have a couple of tests of (relatively) large files. Though they don't have to be actually big, we could just set the chunk size to small, like 200 bytes, and have a 1000 byte file. The main things I'd want to test for are: boundary spanning a chunk, data more than 1 chunk, data more than a few chunks.

benhoyt replied:
It would also be good to add a few tests to `TestRealPebble` and test them against the real Go-based pebble server. Or *at least* do thorough manual testing against a real pebble via the test CLI -- but it might be easier to just add `TestRealPebble` tests anyway.

rwcarlsen replied:
I added a round trip push+pull real pebble test and also one test with a tiny chunk size causing data and boundaries to span multiple chunks.  Caught a small bug actually!

benhoyt commented on ops/pebble.py:
```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                pass
+            else:
+                indices.add(index)
+        return min(indices) if indices else None
```
It seems heavy to use a for loop and a set for this. Also, you can use `buffer.find()` instead of `buffer.index()` to avoid the exception. Also, it seems to me the terminal boundary should never be before the regular boundary, so we could check that. What about something like this:

```
def _get_next_boundary_index(self):
    boundary_index = self._buffer.find(self._boundary)
    terminal_index = self._buffer.find(self._terminal_boudnary)
    if boundary_index >= 0:
        if terminal_index >= 0 and terminal_index < boundary_index:
            raise RuntimeError('terminal boundary found before regular boundary')
        return boundary_index
    if terminal_index >= 0:
        return terminal_index
    return None
```

jnsgruk replied:
Looks like this was done in a later commit :)

benhoyt commented:
Left a bunch of review comments. I think there are a couple of minor things that need fixing, but overall I think this is significantly clearer and simpler than the FSM approach -- nice work!

rwcarlsen commented:
`tempfile.NamedTemporaryFile` says that re-opening a file after closing only works on unix-based systems.  The code as currently written relies on this - do we care about non-unix systems?

Regarding tempfile cleanup - the temp file object supports automatic deletion (when the file is gc'd or closed) - but that doesn't allow the file we return to the caller of `pull` to have different encoding/mode/permissions than those used by the parser when writing the file.

Would it be worth modifying `pull` to optionally take a writeable stream/buffer arg that the parser dumps the file contents into - e.g. 

```python
def pull(self, path: str, dst : io.RawIOBase = None, ...)
    ...

# using pull:
with open('myfile', 'wb') as f:
    container.pull(path, dst=f) # maybe pull does seek(0) on f before returning
    data = f.read()
    ...

# or
buf = container.pull(path, dst=io.StringIO())
data = buf.read()
buf.close()
...
```

seems a little unpythonic (more Go-ey) :-/  Or maybe we just mv the temp file to a non-temp location after parsing is done...

jameinel reviewed (commented):
So one fundamental tweak I want to request, we are currently searching through the buffer 4 times (each _next_boundary is doing 2 searches, and we have 1 call in the core part of feed, and then another call inside of the while True: loop)
I think with a couple small tweaks we can easily drop that to a single search through the buffer on the happy path. (utilize the fact that boundary and terminal boundary share a common prefix, and only search for the next index right before you actively use it)

I'd love a design that doesn't require a temp file that buffers everything, but I think that is out of scope for this PR.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1953,175 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+
+    def feed(self, data: bytes):
+        """Provide more data to the parser to process."""
+        self._buffer.extend(data)
```
This is definitely something we can revisit if performance or memory becomes an issue.  But I think keeping the implementation trivial like it is now is reasonable.  I know @pengale has been ruminating on some profiling processes/tooling for the framework.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1953,175 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        # RFC2046 (p. 19) states that there generally shouldn't be content
+        # before the first boundary and after the last, but there could be and
+        # it should be ignored.
+        next_boundary_index = self._get_next_boundary_index()
```
That code only fires when the while loop will go into the header branch - and it only fires once - then that flag gets set to false and that search is skipped for all remaining calls to feed.  So there aren't actually redundant calls to get the boundary index.

jameinel replied:
The way the code is written, isn't this triggered every time the pull() code calls to feed() ?
Which then it does a second time when it gets to 
https://github.com/canonical/operator/pull/667/files#diff-d74b54ce28b898e8febb0f33f24abc9bf91bb7793c51ae716bf9a9816d90104dR2047-R2048
```
            elif self._part_type == 'files':
                next_boundary_index = self._get_next_boundary_index()
```

hence my statement to move the search inside the if self._pre_first_boundary check.

rwcarlsen replied:
Sorry - I misread your original suggestion.  You are correct.  It's fixed now.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1466,45 +1467,31 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
+        parser = MultipartFileParser(boundary)
 
-        # Then read the rest of the response and feed it to the parser.
         while True:
```
That is a good suggestion.  We do have the name as soon as we read the part header - which precedes the file data - so it should be possible to do something like this.  Perhaps it would be best to create a separate issue for push/pull performance evaluation to facilitate discussion and tracking for these kinds of improvements (see #684).

rwcarlsen commented:
All have been addressed except this one:

How should we handle temp file cleanup on windows?  Our options are:

* re-add a `__del__`/gc tempfile unlink/removal - the problem here is that __del__ was actually on the parser object - but we want the file to stick around until the file-object is gc'd which could be long after the parser object is gc'd - so maybe this isn't an option.
* wrap the file in e.g. a `tempfile._TemporaryFileWrapper` that will auto-delete on close or gc before returning it from `pull`
* do nothing - just let the temp file live on in windows environments/containers.

It's still not clear to me under what circumstances charm code would be running on Windows.

jnsgruk commented:
Nice work, Robert. I think option 2 sounds like the most appropriate here.

benhoyt commented on ops/pebble.py:
```diff
@@ -62,7 +63,7 @@ def __init__(self, host, timeout=_not_provided, socket_path=None):
         """Override connect to use Unix socket (instead of TCP socket)."""
         if not hasattr(socket, 'AF_UNIX'):
-            raise NotImplementedError('Unix sockets not supported on {}'.format(sys.platform))
+            raise NotImplementedError('unix sockets not supported on {}'.format(sys.platform))
```
Nit: I don't think this should be changed -- "Unix" is a proper noun so should be capitalized.

rwcarlsen replied:
fixed.

benhoyt commented on ops/pebble.py:
```diff
@@ -1466,45 +1467,39 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
+        # closed.  This prevents callers from interacting with it on disk.
+        # But e.g. windows doesn't allow removing opened files like this.
+        if os.name != 'posix' or sys.platform == 'cygwin':
+            return tempfile._TemporaryFileWrapper(f, f.name, delete=True)
```
Hmm, I don't love this, though I don't know of a better way offhand.

1) Using `os.name` and `sys.platform` is a bit of a code smell - is there any way to use feature detection instead?
2) `tempfile._TemporaryFileWrapper` is an undocumented, "private" class, so it doesn't seem wise to use it. Then again, I can see why you have, as it's a fair bit of gnarly code. And it's been there for years, so unlikely to go away anytime soon ... as such, I guess depending on it is reasonable.

rwcarlsen replied:
I don't love it too much either.  I thought about this quite a bit - and while it does seem a bit hacky - this is in fact exactly how the tempfile lib discovers how to treat temp file opening/deletion differently for windows. I basically pasted this check from the stdlib.

rwcarlsen replied:
Also - on the undoc'd stdlib function `_TemporaryFileWrapper` - I did want to make sure that that class/interface was relatively dependable - and indeed the API as far as we are using has been stable since well back into python2.  I'm definitely open to alternative suggestions, but this seemed like a reasonable approach.

benhoyt replied:
Fair -- thanks for the due diligence!

benhoyt reviewed (approved):
One nit, one question, but approved overall. (I reviewed the decoding logic earlier.) Thanks!

rwcarlsen commented:
@niemeyer - this has been sitting for a couple weeks.  Are you still planning on taking another look or are we clear to move forward?

niemeyer commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        content_disposition = self._headers.get_content_disposition()
+        if content_disposition != 'form-data':
+            raise RuntimeError(
+                'unexpected content disposition: {}'.format(content_disposition))
```
Similar to earlier comment, let's use the repr when dealing with bad data as we don't know what kind of garbage made it invalid.

Also, this should not be a RuntimeError. We're parsing some data from the network, and got a value we don't expect. This isn't an error in the local runtime.

I believe we're using ProtocolError for that kind of issue in this file, but please have a look.  There are other cases like this that need fixing too. Would you mind to have a look please?

rwcarlsen replied:
Switched all runtime errors to protocol errors with repr.

niemeyer commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        elif name == 'files':
+            filename = self._headers.get_filename()
+            if filename is None:
+                raise NotImplementedError('multipart "files" part missing filename')
```
That doesn't look like a NotImplementedError either, I think?  Isn't this typically used for Python methods/etc that are missing/abstract?

niemeyer commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+    def _prepare_for_part_body(self):
+        name = self._headers.get_param('name', header='content-disposition')
+        if name == 'response':
+            part_type = 'response'
```
Why not `part_type = ...` above?  We're manually doing that in all branches of the conditional.

rwcarlsen replied:
done

niemeyer commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        return list(self._files.keys())
+
+    def get_file(self, path, encoding):
+        """Return an open file object containing the data of the specified file."""
```
"containing the data"? Probably a leftover.

rwcarlsen replied:
fixed.

niemeyer reviewed (changes requested):
Thanks again for picking this up, Robert, and sorry for the delay on the response.

It's looking pretty close. I believe after sorting out these last few comments we can get it in:

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1466,45 +1467,41 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
         self._raise_on_path_error(resp, path)
 
-        if content is None:
+        if len(filenames) < 1:
```
`_raise_on_path_error` checks that the json response metadata we received from pebble is consistent with the (multipart) file parts we got from pebble, and that the pebble response didn't have any errors in it; this is semantically a bit different from the other checks.  However, the other checks are a little inconsistent and sloppy - I'll tidy them up a bit.

rwcarlsen replied:
cleaned up.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1466,45 +1467,41 @@ def pull(self, path: str, *, encoding: str = 'utf-8') -> typing.Union[typing.Bin
+        if filenames:
+            filename = filenames[0]
+            if filename != path:
+                raise ProtocolError('path not expected: {}'.format(filename))
```
done

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
         self._request('POST', '/v1/signals', body=body)
+
+
+class MultipartFileParser:
```
Agree.  Done.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        self._pre_first_boundary = True
+
+    def remove_files(self):
+        """remove/unlink all temporary files stored on disk that were parsed from the feed."""
```
done

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                self._buffer = self._buffer[next_boundary_index:]
+
+        while True:
+            # Try to read headers, if present.
```
done.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    self._headers = None
+                else:
+                    # Insufficient data to continue
+                    break
```
I added an 8 MB limit.  In practice, I think the headers and response part size shouldn't be getting very large at all.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to the next boundary line.
+                    safe_bound = max(0, len(self._buffer) - len(self._terminal_boundary))
```
conditional added.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    # Advance the buffer for bytes consumed
+                    self._buffer = self._buffer[safe_bound:]
+
+                    # Insufficient data to continue
```
done

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        if self._buffer[boundary_index:].startswith(self._boundary):
+            return boundary_index
+        elif self._buffer[boundary_index:].startswith(self._terminal_boundary):
+            return boundary_index
```
I ended up needing to rewrite this function to handle the linear whitespace.  It should be better now.

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1966,3 +1963,177 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        if self._pre_first_boundary:
+            next_boundary_index = self._get_next_boundary_index()
+            if next_boundary_index is None:
+                return
```
fixed.

rwcarlsen commented:
This should be ready.

rwcarlsen commented:
@niemeyer's comments helped me see that the parser as coded was a bit tricky/unclear to follow.  I reworked it a bit - separating the protocol/pebble message bits away from the mime multipart bits - and I think things are much more clear.  I also added several table-driven tests for the multipart parsing giving me better confidence that things are working like we want.

niemeyer commented:
@jnsgruk @rwcarlsen Five stars for caring about the layout of the code and the design of the parser, but this seems to have completely redesigned the logic from the ground up once more, which means it now needs a proper review once more too. Can you please ask for someone else (Ben? Harry?) to review this logic first so you're not holding on for me right now, as I'm traveling and will be on another sprint in the next few days?

benhoyt commented on ops/pebble.py:
```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        return open(self._files[path].name, mode, encoding=encoding, newline=newline)
+
+
+def _multipart_parser(
```
Is this done as a separate function to make unit testing easier? I suppose that's fair enough, however, it seems to me it'd be simpler to just make these methods part of the main `_FilesParser` class, to avoid having to pass in the `handle_*` functions. It would just call `self.handle_*` and update its own state. That said, I don't feel strongly about this if the separation makes testing easier (assuming we're also testing the rest of the `_FilesParser` methods).

benhoyt commented on ops/pebble.py:
```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        max_boundary_length = len(b'\r\n--' + marker + b'--\r\n') + 99
+
+    def feed(data: bytes):
+        nonlocal buf, pos, done
```
This pattern of returning a stateful function seems highly unusual to me in Python. I've very rarely seen `nonlocal` used. I think this would be clearer and more idiomatic as just a `_MultipartParser` class with the `nonlocal` variables just as instance variables. It would not be any more code, just an `__init__` and a `feed` method.

benhoyt commented on ops/pebble.py:
```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                # parse the part header
+                if max_lookahead and len(buf) - pos > max_lookahead:
+                    raise ProtocolError('header terminator not found')
+                if header_terminator not in buf:
```
Probably better to use `buf.find(header_terminator)` and assign it to a variable, to avoid the search twice (once here and once below with `buf.index`).

rwcarlsen replied:
done

benhoyt commented on test/test_pebble.py:
```diff
@@ -893,6 +894,134 @@ def build_mock_change_dict(change_id='70'):
 
 
+class TestMultipartParser(unittest.TestCase):
+    class _Case:
```
I was thinking a `collections.namedtuple` would be simpler for this, but I guess it's tricky to do a default (for `max_boundary`) with a namedtuple...

rwcarlsen replied:
Yeah - I don't feel strongly any particular way here, but the class does give better options for defaults and such.

benhoyt commented on test/test_pebble.py:
```diff
@@ -893,6 +894,134 @@ def build_mock_change_dict(change_id='70'):
+                b'\r\n--qwerty \t \r\nheader foo\r\n\r\nfoo bar\r\n--qwerty--\r\nheader bar\r\n\r\nfoo baz\r\n--qwerty--\r\n',  # noqa
+                [b'header foo\r\n\r\n'],
+                [b'foo bar'],
+                want_bodies_done=[True],
```
These look like good cases. Have you looked at the code coverage to see we're covering all the cases/branches in the code under test?

rwcarlsen replied:
Looks like there are 2 uncovered lines in entire multipart parser - the header terminator not found error branch, and the max boundary length default - which is always set explicitly in the table tests.

rwcarlsen replied:
I added a case to check for the missing header error.

benhoyt commented on test/test_pebble.py:
```diff
@@ -893,6 +894,134 @@ def build_mock_change_dict(change_id='70'):
+                    bodies[-1] += data
+                    bodies_done[-1] = done
+
+                p = pebble._multipart_parser(
```
Nit: `p` -> `parser` (a single-char var name seems extreme here)

rwcarlsen replied:
done

benhoyt commented on test/test_pebble.py:
```diff
@@ -2775,6 +2946,16 @@ def test_exec_send_stdin(self):
+        with self.client.pull(fname) as f:
+            data = f.read()
+            self.assertEqual(data, content)
+        process = self.client.exec(['/bin/rm', fname])
```
Guess we could just use `os.remove` here, as it's always on the same machine?

rwcarlsen replied:
changed.

benhoyt reviewed (commented):
Looks good, and definitely a clearer structure. A few comments, only a couple of them non-nits.

Also, have you done a bit of manual testing with a few large files (binary and text) and ensure they round-trip correctly? (Eg: same sha1sum before and after)

jnsgruk commented on ops/pebble.py:
```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+class _PebbleFilesParser:
+    """A limited purpose multi-part parser backed by files for memory efficiency."""
+
+    def __init__(self, boundary: bytes):
```
+1 on using something like `mypy` or `pyright` - probably the latter due to the significantly more helpful output - perhaps one for a separate issue though!

PietroPasotti replied:
I like type annotations even without type checkers. They help write well-documented code and if your IDE supports it, they help spot simple bugs early on and they even help with code completion in some cases (Literals).

rwcarlsen replied:
fixed. (no mypy or pyright yet)

PietroPasotti commented on ops/pebble.py:
```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    handle_body(data)
+                    return  # waiting for more data
+                else:
+                    return  # waiting for mor data
```
typo: more

rwcarlsen replied:
fixed

rwcarlsen commented on ops/pebble.py:
```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+    r"""Returns a stateful function for parsing mime multipart messages.
+
+    Args:
+        marker: the multipar boundary marker (i.e. in "\r\n--<marker>--\r\n")
```
fixed

jnsgruk commented:
@niemeyer this is ready for another pass now I believe :)
